AG-9272 QA Feedback: Click Events

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-9272

olegat committed:
AG-9272 Handle nodeClick and nodeDoubleClick events

olegat committed:
AG-9272 Avoid firing (dbl)click events twice

olegat committed:
AG-9272 Handle nodeClickRange in errorBar.ts

This hit-testing for the highlight/tooltips uses a different range than
click events (`tooltip.range` versus `series.nodeClickRange`).

This commits adds the ability for the error bar module to perform hit
testing for both range values.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -297,25 +303,89 @@ export class ErrorBars
+        // We don't want that because there is an API for adding listeners to 'nodeClick' and
+        // 'nodeDoubleClick' events. Therefore, check whether the event hits a series node datum:
+        if (range === 'exact' || range === 0) {
+            const seriesNode = this.cartesianSeries.contentGroup.pickNode(event.offsetX, event.offsetY);
```
`pickNode()` is a potentially expensive operation, we should avoid executing it multiple times (it's already in-play from a series perspective?) here and below.

Maybe we could consider 'disarming' the error-bar clickable state based upon the currently highlighted datum state?

olegat replied:
Yes, `pickNode()` is not ideal. However, we cannot use the highlight state to determine whether a node is clickable because we provide the option to customise the highlight and click range separately:
```
{
  tooltip: { range: "exact" },
  series: [ { nodeClickRange: 10 } ],
}
```
This example makes it possible to click an unhighlighted node if the user clicks within a range of 10 pixels.

Essentially this function just needs to check whether the series node datum that the error bars map to is handling the event. So it doesn't need to do hit-testing for all of the nodes in `contentGroup`. We might be able to add a optimised method to just do hit-testing on a single a datum. Example:
```typescript
private checkInteractionEventOverlap(datum: ErrorBarNodeDatum, event: InteractionEvent, range: InteractionRange): boolean {
    if (this.cartesianSeries.checkDatumHit(datum, event, range)) {
        return false;
    }
    return true;
}
```

Ideally, we want either the series datum node or the error bar node to consume the click event to avoid firing the `nodeClick` event twice. There is an `InteractionEvent.consume()` function. I tried calling that function in the _errorBar.ts_ file but it didn't prevent both nodes from firing the `nodeClick` event.

olegat replied:
Okay, so I've discovered that the `InteractionEvent.consume()` method is intended to suppress the event to other "internal" listeners like the legend and the error bars modules. The Series class could in theory consume this to hide it from all the other listener, but I don't think that this is desirable (it might unintentionally break things with the legend, for example).

Instead, I've added the `InteractionEvent.callbackComplete` property to keep track of whether or not the nodeClick / nodeDoubleClick have been broadcasted to the library user. https://github.com/ag-grid/ag-charts/commit/8193d3cdc36f2308da8121e1937069fc9b6aa8a6

I think that this is the best solution, because it avoids the need to recalculate the hit-tests and keeps a clear distinction between internal & external listeners.

olegat replied:
This has been changed. `ErrorBars` now implements `pickNode` for exact, nearest and main-axis-first pick modes. PTAL.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBarNode.ts:
```diff
@@ -170,52 +200,57 @@ export class ErrorBarNode extends _Scene.Group {
         caps.markDirtyTransform();
+    }
+
+    updateBBoxes(...ranges: InteractionRange[]): void {
```
I'm just wondering if we should improve the scene-graph support for this type of operation, rather than have all of this behaviour as `ErrorBar`-specific functionality?

`Group.computeBBox()` already exists, and I think we could improve the `Group.containsPoint()` handling for this case (specifically, allow `Group` to configurably have accurate bounding-box handling)?

olegat replied:
I agree that this should be a built-in scene graph feature. However, adding such a feature to the scene has the potential of causing many bug and performance regressions across the entire ag-charts framework.

Are we confident enough with our test coverage to mitigate this risk so close the initial release? Or should we leave this in the backlog?

olegat replied:
Some this has been moved into the community package (new file _nearest.ts_), but the `ErrorBarNode` still cares take of calculating the size & position of bounding boxes. PTAL.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBarNode.ts:
```diff
@@ -29,17 +30,46 @@ export type ErrorBarCapFormatter = (params: AgErrorBarFormatterParams) => AgErro
+    return typeof range === 'string' ? 0 : range;
+}
+
+class HierarchialBBox {
```
See my other comment, but I'm thinking these behaviours are not error-bar specific?

olegat replied:
We can safely make `HierarchialBBox` part of the ag-chart-community package as a utility class. Although I don't see much benefit to this. The `HierarchialBBox` class only supports a tree of depth 1, so it's not such a useful utility class. A more useful version of `HierarchialBBox` would support trees with multiple levels:
```diff
 class HierarchialBBox {
     private readonly union: BBox;
-    private readonly components: BBox[];
+    private readonly components: HierarchialBBox[];
     ...
 }
```

olegat replied:
What shall we do with this `HierarchialBBox` class? I can move it into the community package as a scene utility and have `ErrorBarNode` extend it implement hit testing. Please let me know what you think.

olegat committed:
AG-9272 CR: Add callbackComplete flag to events

The errorBar.ts needs to know whether the nodeClick/nodeDoubleClick event
has already been fired by the series to broadcasting this to the user
twice. This is what the callbackComplete flag is intended for. This is
better for performance, because it avoid the need are redoing the
hit-testing.

olegat committed:
AG-9272 Implement pickNode in errorBar.ts

After some discussion, we decided not to mutate the InteractionEvent.
Instead, errorBar.ts just implements hit-testing for exact, nearest
and mainAxisFirst pick modes.

The CartesianSeries class also loops through the modules to determine
which node is the nearest. With this, the ErrorBars class no longer
needs to listen for input events because that's handled by the Chart
class.

alantreadway commented on packages/ag-charts-community/src/module/moduleMap.ts:
```diff
@@ -55,4 +55,10 @@ export class ModuleMap<M extends Module<C, I>, C, I extends ModuleInstance = Mod
+
+    *[Symbol.iterator](): IterableIterator<I> {
+        for (const { instance } of Object.values(this.modules)) {
+            yield instance;
```
Interesting, this is the first practical use of `yield` I've seen in a codebase :D

olegat replied:
I actually think that it might make more sense for `Series` to extend `ModuleMap` and have `this.modules` be protected, but that's a discussion for another time 😄

olegat committed:
AG-9272 Fix eslint errors

alantreadway reviewed (approved):
LGTM
